specs/lightclient/beacon-chain.md

```diff
@@ -703,79 +655,68 @@ def process_justification_and_finalization(state: BeaconState) -> None:
         state.finalized_checkpoint = old_current_justified_checkpoint
 ```
 
+#### Leak updates
+
+*Note*: The function `process_leak_updates` is new.
+
+```python
+def process_leak_updates(state: BeaconState) -> None:
+    # In a given epoch the leak score of a validator:
+    #    (a) decreases by 1 if the validator participates
+    #    (b) increases by LEAK_SCORE_BIAS if is_leak_active(state) and the validator does not participate
+    # For efficiency rule (b) increases are batched at activation-exit epochs using the leak epochs counter.
+
+    # Decrease leak score per rule (a)
+    decrease_amount = 1
+    if is_leak_active(state):
+        decrease_amount += LEAK_SCORE_BIAS  # Cancel corresponding increase at activation-exit epochs
+    for index in get_unslashed_participating_indices(state, TIMELY_TARGET_FLAG_INDEX, get_previous_epoch(state)):
+        if state.leak_scores[index] < decrease_amount:
+            state.leak_scores[index] = 0
+        else:
+            state.leak_scores[index] -= decrease_amount
+
+    # Increase leak score per rule (b)
+    if is_leak_active(state):
+        state.leak_epochs_counter += Epoch(1)
+    if is_activation_exit_epoch(state):
+        for index in get_eligible_validator_indices(state):
+            state.leak_scores[index] += LEAK_SCORE_BIAS * state.leak_epochs_counter
+        state.leak_epochs_counter = Epoch(0)
+```
+
 #### Rewards and penalties
 
 *Note*: The function `process_rewards_and_penalties` is modified to support the incentive reforms.
 
+
 ```python
 def process_rewards_and_penalties(state: BeaconState) -> None:
     # No rewards are applied at the end of `GENESIS_EPOCH` because rewards are for work done in the previous epoch
     if get_current_epoch(state) == GENESIS_EPOCH:
         return
 
-    process_rewards(state)
-    process_penalties(state)
```

Comment:
I can see this changes make sense, but IMHO it's good to have multiple purpose-focused functions. It would be more flexible when we have to modify it in the future forks.

Reply:
yeah, I would tend to agree